### PR TITLE
feat: support bn254 precompile

### PR DESCRIPTION
**Pull request by beqaabu**
# Description of proposed changes

<describe what this PR is about and why we want it>

---

Practical tips for PR review & merge:

- [ ] All GitHub Actions report PASS
- [ ] Newly added code/functions have unit tests
  - [ ] Coverage tools report all newly added lines as covered
  - [ ] The positive scenario is exercised
  - [ ] Negative scenarios are exercised, e.g. assert on all possible errors
  - [ ] Assert on events triggered if applicable
  - [ ] Assert on changes made to storage if applicable
- [ ] Modified behavior/functions - try to make sure above test items are covered
- [ ] Integration tests are added if applicable/needed

### CODE REVIEW

**Review by Trantorian1 — DISMISSED** [review]

File: runtime/src/precompiles.rs [hunk 1]
```diff
@@ -45,6 +50,9 @@ where
             a if a == hash(3) => Some(Ripemd160::execute(handle)),
             a if a == hash(4) => Some(Identity::execute(handle)),
             a if a == hash(5) => Some(Modexp::execute(handle)),
+            a if a == hash(6) => Some(Bn128Add::execute(handle)),
+            a if a == hash(7) => Some(Bn128Mul::execute(handle)),
+            a if a == hash(8) => Some(Bn128Pairing::execute(handle)),
```
Comment (atodorov) [inline runtime/src/precompiles.rs:57]: These 3 need some tests. 

Inside creditcoin3-next see: cli/src/test/blockchain-tests/precompiles/.

Also see `substrate-transfer/src/tests.rs` - you may want to add only unit tests that will call these new precompiles and cover basic positive/negative scenarios. (I'm not sure how well these are tested upstream but we should be exercising them with the inputs we deem important). 


NOTE: it could prove easier to first add this into creditcoin3-next and then port either all of it or only the rust bits instead of the other way around.

**Review by atodorov — CHANGES_REQUESTED** [review]
Missing test coverage, see notes.

**Review by atodorov — APPROVED** [review]

**Review by atodorov — CHANGES_REQUESTED** [review]
Needs a version bump

**Review by atodorov — APPROVED** [review]

**Review by Trantorian1 — APPROVED** [review]

**Review by didac-gluwa — APPROVED** [review]

File: Cargo.toml [hunk 2]
```diff
@@ -153,7 +153,7 @@ pallet-transaction-payment-rpc = { git = "https://github.com/paritytech/polkadot
 pallet-transaction-payment-rpc-runtime-api = { git = "https://github.com/paritytech/polkadot-sdk", branch = "stable2409", default-features = false }
 pallet-utility = { git = "https://github.com/paritytech/polkadot-sdk", branch = "stable2409", default-features = false }
 # Substrate Utility
-frame-benchmarking-cli = { git = "https://github.com/paritytech/polkadot-sdk", branch = "stable2409" }
+frame-benchmarking-cli = { git = "https://github.com/paritytech/polkadot-sdk", branch = "stable2409", version = "43.1.0" }
```
Comment (DylanVerstraete) [inline Cargo.toml]: why is this required?

Comment (beqaabu) [reply Cargo.toml]: stragely by the time of creating this pr the node didnt used to build unless the version was specified. I checked it now and it works without the version so I'll change it back

**Review by DylanVerstraete — APPROVED** [review]
LGTM 👍

**Review by Trantorian1 — APPROVED** [review]

**Comment by beqaabu** [comment]
see passing hardhat job for more details https://github.com/gluwa/creditcoin3/actions/runs/20176954863/job/57928371622?pr=683, namely "Test contracts" step
